qiskit/synthesis/linear/linear_circuits_utils.py

```diff
@@ -73,30 +91,99 @@ def optimize_cx_4_options(function: Callable, mat: np.ndarray, optimize_count: b
             mat_cpy = calc_inverse_matrix(mat_cpy)
             qc = function(mat_cpy)
             qc = qc.inverse()
+            circuits.append(qc)
         elif i == 2:
             mat_cpy = np.transpose(mat_cpy)
             qc = function(mat_cpy)
             qc = transpose_cx_circ(qc)
+            circuits.append(qc)
         elif i == 3:
             mat_cpy = calc_inverse_matrix(np.transpose(mat_cpy))
             qc = function(mat_cpy)
             qc = transpose_cx_circ(qc)
             qc = qc.inverse()
+            circuits.append(qc)
+
+    return circuits
 
-        new_depth = qc.depth()
-        new_count = qc.count_ops()["cx"]
-        # Prioritize count, and if it has the same count, then also consider depth
-        better_count = (optimize_count and best_count > new_count) or (
-            not optimize_count and best_depth == new_depth and best_count > new_count
-        )
-        # Prioritize depth, and if it has the same depth, then also consider count
-        better_depth = (not optimize_count and best_depth > new_depth) or (
-            optimize_count and best_count == new_count and best_depth > new_depth
-        )
-
-        if better_count or better_depth:
-            best_count = new_count
-            best_depth = new_depth
-            best_qc = qc
 
+def _choose_best_linear_circuit(
+    circuits: List[QuantumCircuit], optimize_count: bool = True
+) -> QuantumCircuit:
+    """Returns the best quantum circuit either in terms of gate count or depth.
+
+    Args:
+        circuits: a list of quantum circuits
+        optimize_count: True if the number of CX gates is optimized, False if the depth is optimized.
+
+    Returns:
+        QuantumCircuit: the best quantum circuit out of the given circuits.
+    """
+    best_qc = circuits[0]
+    for circuit in circuits[1:]:
+        if _compare_linear_circuits(best_qc, circuit, optimize_count=optimize_count):
+            best_qc = circuit
     return best_qc
+
+
+def _linear_circuit_depth(qc: QuantumCircuit):
```

Comment:
It's worth to add a test for this method

Reply:
Thanks! I am not sure how much testing we need for internal functions, but I guess a few tests can't hurt. Done in [a324253](https://github.com/Qiskit/qiskit-terra/pull/9250/commits/a3242535bd04b7cf1395e7624940c283491de09b).